File: easy/array/cpp/238-move-zeroes.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using std::vector;

class Solution
{
public:
    void moveZeroes(vector<int> &nums)
    {
        int slow = 0;
        for (size_t i = 0; i < nums.size(); i++)
        {
            int prev = nums[slow];
            int curr = nums[i];

            if (prev != 0)
            {
                slow += 1;
                continue;
            }

            if (curr != 0 && prev == 0)
            {
                std::iter_swap(nums.begin() + slow, nums.begin() + i);
                slow += 1;
            }
        }
    }
};

File: easy/array/python/238-move-zeroes.py
import enum
from typing import List


class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        slow = 0
        for fast in range(len(nums)):
            curr = nums[fast]
            prev = nums[slow]

            if prev != 0:
                slow += 1

            if curr != 0 and prev == 0:
                nums[fast], nums[slow] = nums[slow], nums[fast]
                slow += 1


nums = [0, 1, 0, 3, 12]
Solution().moveZeroes(nums)
print(nums)
nums = [1, 2, 3, 4]
Solution().moveZeroes(nums)
print(nums)
nums = [1]
Solution().moveZeroes(nums)
print(nums)
nums = [0, 0, 0]
Solution().moveZeroes(nums)
print(nums)
nums = [0, 1, 0]
Solution().moveZeroes(nums)
print(nums)
nums = [0, 1, 2]
Solution().moveZeroes(nums)
print(nums)
nums = [1, 0, 1]
Solution().moveZeroes(nums)
print(nums)
